test_conformance/api/negative_device.cpp

```diff
@@ -0,0 +1,520 @@
+    if (supported_properties & SupportedPartitionSchemes::Equally)
+    {
+        properties[0] = CL_DEVICE_PARTITION_EQUALLY;
+        properties[1] = 1;
+        properties[2] = 0;
+        out_devices.resize(static_cast<size_t>(max_compute_units));
+        max_for_partition = max_compute_units;
+    }
+    else if (supported_properties & SupportedPartitionSchemes::Counts)
+    {
+        properties[0] = CL_DEVICE_PARTITION_BY_COUNTS;
+        properties[1] = 1;
+        properties[2] = CL_DEVICE_PARTITION_BY_COUNTS_LIST_END;
+        out_devices.resize(static_cast<size_t>(max_sub_devices));
+        max_for_partition = max_sub_devices;
+    }
+    else
+    {
+        properties[0] = CL_DEVICE_PARTITION_BY_AFFINITY_DOMAIN;
+        properties[1] = CL_DEVICE_AFFINITY_DOMAIN_NEXT_PARTITIONABLE;
+        properties[2] = 0;
+    }
+
+    properties[3] = 0;
+
+    cl_int err(CL_SUCCESS);
+    for (auto invalid_device : get_invalid_objects<cl_device_id>(deviceID))
+    {
+        err = clCreateSubDevices(invalid_device, properties, 1,
+                                 out_devices.data(), nullptr);
+        test_failure_error_ret(err, CL_INVALID_DEVICE,
+                               "clCreateSubDevices should return "
+                               "CL_INVALID_DEVICE when: \"in_device "
+                               "is not a valid device\"",
+                               TEST_FAIL);
+    }
+
+    err = clCreateSubDevices(deviceID, nullptr, 1, out_devices.data(), nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clCreateSubDevices should return CL_INVALID_VALUE when: \"values "
+        "specified in properties are not valid\" using a nullptr",
+        TEST_FAIL);
+
+    err = clCreateSubDevices(deviceID, properties, 0, out_devices.data(),
+                             nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clCreateSubDevices should return CL_INVALID_VALUE when: \"out_devices "
+        "is not NULL and num_devices is less than the number of sub-devices "
+        "created by the partition scheme\"",
+        TEST_FAIL);
+
+    if (supported_properties != SupportedPartitionSchemes::All_Schemes)
+    {
+        std::vector<cl_device_partition_property> invalid_properties =
+            get_invalid_properties(supported_properties
+                                   ^ SupportedPartitionSchemes::All_Schemes);
+        err = clCreateSubDevices(deviceID, invalid_properties.data(), 1,
+                                 out_devices.data(), nullptr);
+        test_failure_error_ret(
+            err, CL_INVALID_VALUE,
+            "clCreateSubDevices should return CL_INVALID_VALUE when: \"values "
+            "specified in properties are valid but not supported by the "
+            "device\"",
+            TEST_FAIL);
+    }
+
+    err = clCreateSubDevices(deviceID, properties, max_for_partition + 1,
+                             out_devices.data(), nullptr);
```

Comment:
I think there is still a problem here.  On the Intel CPU OpenCL implementation, testing on a CPU with 24 cores, I am seeing this call being generated:

```
>>>> clCreateSubDevices: in_device = 12th Gen Intel(R) Core(TM) i9-12900K (CL_DEVICE_TYPE_CPU) (0x5e7b00958778), properties = [ CL_DEVICE_PARTITION_EQUALLY = 1 ], num_devices = 25
<<<< clCreateSubDevices -> CL_SUCCESS
```

This call is returning `CL_SUCCESS`, and I don't see any reason why it should not return `CL_SUCCESS`: it is only creating 24 sub-devices with this call and the 25th is un-written, but this should be fine.  This certainly is not testing the case where "in_device could not be further partitioned".

Perhaps the properties should be different to test this case?  For example, I believe trying to partition equally into a device with 24 compute units will generate the expected error, since in this case the input device is not being further partitioned:

```
>>>> clCreateSubDevices: in_device = 12th Gen Intel(R) Core(TM) i9-12900K (CL_DEVICE_TYPE_CPU) (0x5f50d1d32778), properties = [ CL_DEVICE_PARTITION_EQUALLY = 24 ], num_devices = 1
ERROR! clCreateSubDevices returned CL_DEVICE_PARTITION_FAILED (-18)
<<<< clCreateSubDevices -> CL_DEVICE_PARTITION_FAILED
```